# Coding Exercises (Appendix 3)

Now, you will have the opportunity to practice what you have learned. <br>
__Follow the instructions__ and insert your code! 

The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need some help or if you want to check your code, you can also have a look at the __solutions__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Exercise 8: N-Dimensional Numpy Arrays

In [2]:
# run the cell!
import numpy as np
np.set_printoptions(precision=2, suppress= True)

Let´s assume we have three investment projects with the following cashflow plans:

In [3]:
# run the cell!
project_1 = [-500, 100, 150, 200, 200, 175, 100, 50]
project_2 = [-750, 100, 125, 150, 175, 200, 250, 0]
project_3 = [-2000, 500, 1000, 2500, 0, 0, 0, 0]

1. Create and save a __2-dimensional Numpy Array__ (in the variable __projects__).

In [5]:
cf = np.array([project_1, project_2, project_3])
cf

array([[ -500,   100,   150,   200,   200,   175,   100,    50],
       [ -750,   100,   125,   150,   175,   200,   250,     0],
       [-2000,   500,  1000,  2500,     0,     0,     0,     0]])

2. Check the __shape__ of projects. __How many rows/columns__ do we have?

In [6]:
cf.shape

(3, 8)

(result: 3 rows, 8 columns)

3. __Slice__ projects for __the second project__.

In [10]:
cf_1 = cf[:, :2]
cf_1
cf_2 = cf[1]
cf_2

array([-750,  100,  125,  150,  175,  200,  250,    0])

4. Return the __initial Investment__ of __Project 3__.

In [12]:
cf_3 = cf[2][0]
cf_3

-2000

(result: -2000)

5. Return the __final cashflow__ in year 7 for __all projects__.

In [23]:
cf

array([[ -500,   100,   150,   200,   200,   175,   100,    50],
       [ -750,   100,   125,   150,   175,   200,   250,     0],
       [-2000,   500,  1000,  2500,     0,     0,     0,     0]])

In [24]:
cf_f = np.sum(cf, axis = 1)
cf_f

array([ 475,  250, 2000])

6. Return the first positive cashflows (__Year 1__) for __Projects 1 and 2__.

In [34]:
cf_p = cf[:2, 1]
cf_p

array([100, 100])

(result: array([100, 100]))

7. __Change__ the initial investment for __project 3__ to __-1900__.

In [35]:
cf[2][0] = -1900
cf

array([[ -500,   100,   150,   200,   200,   175,   100,    50],
       [ -750,   100,   125,   150,   175,   200,   250,     0],
       [-1900,   500,  1000,  2500,     0,     0,     0,     0]])

8. Change the __last cashflows__ in year 7 for all projects to __25__.

In [49]:
cf[:, -1] = 25
cf

array([[ -500,   100,   150,   200,   200,   175,   100,    25],
       [ -750,   100,   125,   150,   175,   200,   250,    25],
       [-1900,   500,  1000,  2500,     0,     0,     0,    25]])

9. Calculate the total cashflows on each timestamp over all three projects (__column-wise sum__).

In [40]:
cf_sum = np.sum(cf, axis=0) # 0열의 총합, 1하면 행의 총합

array([-3150,   700,  1275,  2850,   375,   375,   350,    75])

(results: array([-3150,   700,  1275,  2850,   375,   375,   350,    75]))

10. Calculate __cumulative cashflows__ for all three projects.

In [50]:
cf_cum = np.cumsum(cf, axis=0)
cf_cum

array([[ -500,   100,   150,   200,   200,   175,   100,    25],
       [-1250,   200,   275,   350,   375,   375,   350,    50],
       [-3150,   700,  1275,  2850,   375,   375,   350,    75]])

11. Convert __all projects__ to EUR currency (__multiply with 0.9__).

In [51]:
cf_cum * 0.9

array([[ -450. ,    90. ,   135. ,   180. ,   180. ,   157.5,    90. ,
           22.5],
       [-1125. ,   180. ,   247.5,   315. ,   337.5,   337.5,   315. ,
           45. ],
       [-2835. ,   630. ,  1147.5,  2565. ,   337.5,   337.5,   315. ,
           67.5]])

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Solutions (Stop here if you want to code on your own!)

# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

In [ ]:
# run the cell!
import numpy as np
np.set_printoptions(precision=2, suppress= True)

Let´s assume we have three investment projects with the following cashflow plans:

In [ ]:
# run the cell!
project_1 = [-500, 100, 150, 200, 200, 175, 100, 50]
project_2 = [-750, 100, 125, 150, 175, 200, 250, 0]
project_3 = [-2000, 500, 1000, 2500, 0, 0, 0, 0]

1. Create and save a __2-dimensional Numpy Array__ (in the variable __projects__).

In [ ]:
projects = np.array([project_1, project_2, project_3])

In [ ]:
projects

2. Check the __shape__ of projects. __How many rows/columns__ do we have?

In [ ]:
projects.shape

(result: 3 rows, 8 columns)

3. __Slice__ projects for __the second project__.

In [ ]:
projects[1]

4. Return the __initial Investment__ of __Project 3__.

In [ ]:
projects[-1, 0]

(result: -2000)

5. Return the __final cashflow__ in year 7 for __all projects__.

In [ ]:
projects[:, -1]

6. Return the first positive cashflows (__Year 1__) for __Projects 1 and 2__.

In [ ]:
projects[:2, 1]

(result: array([100, 100]))

7. __Change__ the initial investment for __project 3__ to __-1900__.

In [ ]:
projects[-1, 0] = -1900

In [ ]:
projects

8. Change the __last cashflows__ in year 7 for all projects to __25__.

In [ ]:
projects[:, -1] = 25

In [ ]:
projects

9. Calculate the total cashflows on each timestamp over all three projects (__column-wise sum__).

In [ ]:
projects.sum(axis = 0)

(results: array([-3150,   700,  1275,  2850,   375,   375,   350,    75]))

10. Calculate __cumulative cashflows__ for all three projects.

In [ ]:
projects.cumsum(axis = 1)

11. Convert __all projects__ to EUR currency (__multiply with 0.9__).

In [ ]:
projects * 0.9